packages/input-text/src/TextInput.js

```diff
@@ -0,0 +1,165 @@
+import * as React from "react";
+import PropTypes from "prop-types";
+import Tooltip from "@crave/farmblocks-tooltip";
+import Image, { badgeSizes } from "@crave/farmblocks-image";
+
+import errorIconSrc from "./constants/errorIcon";
+import StyledLabel from "./styledComponents/TextInput";
+
+class TextInput extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      value: this.props.value,
+      invalid: this.props.invalid,
+      showTooltip: false
+    };
+
+    this.onChange = this.onChange.bind(this);
+    this.onInvalid = this.onInvalid.bind(this);
+    this.onMouseOver = this.onMouseOver.bind(this);
+    this.onMouseLeave = this.onMouseLeave.bind(this);
+  }
+
+  render() {
+    const {
+      label,
+      value,
+      onChange,
+      invalid,
+      validationErrors,
+      onInvalid,
+      tooltipText,
+      errorIconSrc,
+      onMouseOver,
+      onMouseLeave,
+      ...otherProps
+    } = this.props;
+    const containerProps = {
+      filled: this.state.value.length > 0,
+      invalid: this.state.invalid,
+      disabled: otherProps.disabled
+    };
+    const inputProps = {
+      value: this.state.value,
+      onChange: this.onChange,
+      onInvalid: this.onInvalid,
+      onMouseOver: this.onMouseOver,
+      onMouseLeave: this.onMouseLeave,
+      ...otherProps
+    };
+    return (
+      <StyledLabel {...containerProps}>
+        {this._renderInput(inputProps)}
+        {this._renderLabel(label)}
+        {this._renderTooltip(this.state.showTooltip, tooltipText)}
+        {this._renderFailedMessages(
+          this.state.invalid,
+          validationErrors,
+          errorIconSrc
+        )}
+      </StyledLabel>
+    );
+  }
+
+  _renderLabel(label) {
+    return label && <div className="label">{label}</div>;
+  }
+
+  _renderInput(inputProps) {
+    return <input type="text" {...inputProps} />;
+  }
+
+  _renderTooltip(visible, text) {
+    return (
+      visible && (
+        <div className="tooltip">
+          <Tooltip text={text} />
+        </div>
+      )
+    );
+  }
+
+  _renderFailedMessages(invalid, errors, iconSrc) {
+    return (
+      invalid && (
+        <div className="messages">
+          {errors.map((text, index) => (
+            <div className="message" key={`err-${index}`}>
+              <Image
+                className="icon"
+                src={iconSrc}
+                badge
+                size={badgeSizes.SMALL}
+              />
+              {text}
+            </div>
+          ))}
+        </div>
+      )
+    );
+  }
+  onChange(event) {
+    this.setState({
+      value: event.target.value,
+      invalid: false
+    });
+    return this.props.onChange(event);
+  }
+
+  onInvalid(event) {
+    event.preventDefault();
+    this.setState({ invalid: true });
```

Comment:
We must keep in mind that the invalid state of the component will be handled by the wg-api (on our case).
This way we must explicitly control the `invalid` prop with new props passed.

What happens if a field is mounted with valid status, then we make some changes and make an api call, this api call returns an error. How do we handle this situation?

Reply:
there is a boolean invalid property, webapp can change that property to true upon receiving the errors from the backend

Reply:
this default wrrapper is for the component to be able to flag as invalid empty fields before submitting (html 5 required attribute)

Reply:
and to override the default native balloon from the browsers

Reply:
Here's the situation that I've commented above
https://stackblitz.com/edit/react-z67acp?file=index.js